### ISSUE TITLE
Hot reload doesn't support single files, breaks container permissions

### ISSUE THREAD

**Issue by Eronarn** [garden-io/garden#1048]
### Current Behavior

I've been testing out the behavior differences between standard and hot reloads. Overall the system works well, but I ran into some painful issues with a Rails application.

My first attempt was to copy over just necessary files and directories, like so:

```
    - source: Procfile.dev
      target: /code/Procfile.dev
    - source: app
      target: /code/app
    - source: bin
      target: /code/bin
    - source: config
      target: /code/config
    - source: lib
      target: /code/lib
    - source: public
      target: /code/public
```

This resulted in a failed sync (and some confusing errors due to that). Here's the `kubectl describe` of the init container:

```
Init Containers:
  garden-sync-init:
    Container ID:  docker://8f8c2ac352e59e38ddbf0df16114a39295b20b8217b08bb8b7e53542365f8d73
    Image:         127.0.0.1:5000/staging/qf:v-21ce1a8fe1
    Image ID:      docker-pullable://127.0.0.1:5000/staging/qf@sha256:008104de41910750a33448c6374b81e35a1944adbb94968ebc34fe51235aae10
    Port:          <none>
    Host Port:     <none>
    Command:
      /bin/sh
      -c
      mkdir -p /.garden/hot_reload/code && cp -r /code/Procfile.dev/ /.garden/hot_reload/code/Procfile.dev/ && mkdir -p /.garden/hot_reload/code && cp -r /code/app/ /.garden/hot_reload/code/app/ && mkdir -p /.garden/hot_reload/code && cp -r /code/bin/ /.garden/hot_reload/code/bin/ && mkdir -p /.garden/hot_reload/code && cp -r /code/config/ /.garden/hot_reload/code/config/ && mkdir -p /.garden/hot_reload/code && cp -r /code/lib/ /.garden/hot_reload/code/lib/ && mkdir -p /.garden/hot_reload/code && cp -r /code/public/ /.garden/hot_reload/code/public/
    State:          Terminated
      Reason:       Error
      Exit Code:    1
      Started:      Mon, 29 Jul 2019 16:36:27 -0400
      Finished:     Mon, 29 Jul 2019 16:36:27 -0400
    Last State:     Terminated
      Reason:       Error
      Exit Code:    1
      Started:      Mon, 29 Jul 2019 16:35:59 -0400
      Finished:     Mon, 29 Jul 2019 16:35:59 -0400
    Ready:          False
    Restart Count:  3
```

In this case, a trailing slash has been appended to the `Procfile.dev`, which I think is what was causing the command to fail (hard to tell because there didn't seem to be log messages from this init container). At least locally, that `cp` syntax is invalid:

```
[eronarn@ip-192-168-10-243 qf]$ cp -r /code/Procfile.dev/ ./f
cp: /code/Procfile.dev/: No such file or directory
```

I read closer through [the documentation](https://docs.garden.io/reference/module-types/container#hotreload) and it looks like there's some inconsistency on whether a single file can be synced, instead of an entire directory:

> hotReload
> Specifies which **files or directories** to sync

> hotReload > sync > source
> POSIX-style path of **the directory** to sync to the target

I next tried syncing everything in the Git repo, since the sync is fairly fast. But the Rails app didn't like that:

```
emitting WebpackAssetsManifest✖ ｢wdm｣: Error: EACCES: permission denied, mkdir '/code/public/packs'
```

```
20:41:10 web.1     | /home/qf/.rvm/rubies/ruby-2.5.3/lib/ruby/2.5.0/fileutils.rb:232:in `mkdir': Permission denied @ dir_s_mkdir - /code/tmp (Errno::EACCES)
20:41:10 web.1     | 	from /home/qf/.rvm/rubies/ruby-2.5.3/lib/ruby/2.5.0/fileutils.rb:232:in `fu_mkdir'
20:41:10 web.1     | 	from /home/qf/.rvm/rubies/ruby-2.5.3/lib/ruby/2.5.0/fileutils.rb:210:in `block (2 levels) in mkdir_p'
20:41:10 web.1     | 	from /home/qf/.rvm/rubies/ruby-2.5.3/lib/ruby/2.5.0/fileutils.rb:208:in `reverse_each'
20:41:10 web.1     | 	from /home/qf/.rvm/rubies/ruby-2.5.3/lib/ruby/2.5.0/fileutils.rb:208:in `block in mkdir_p'
20:41:10 web.1     | 	from /home/qf/.rvm/rubies/ruby-2.5.3/lib/ruby/2.5.0/fileutils.rb:193:in `each'
20:41:10 web.1     | 	from /home/qf/.rvm/rubies/ruby-2.5.3/lib/ruby/2.5.0/fileutils.rb:193:in `mkdir_p'
20:41:10 web.1     | 	from /home/qf/.rvm/gems/ruby-2.5.3@qfr5/gems/railties-5.1.6.2/lib/rails/commands/server/server_command.rb:75:in `block in create_tmp_directories'
20:41:10 web.1     | 	from /home/qf/.rvm/gems/ruby-2.5.3@qfr5/gems/railties-5.1.6.2/lib/rails/commands/server/server_command.rb:74:in `each'
20:41:10 web.1     | 	from /home/qf/.rvm/gems/ruby-2.5.3@qfr5/gems/railties-5.1.6.2/lib/rails/commands/server/server_command.rb:74:in `create_tmp_directories'
20:41:10 web.1     | 	from /home/qf/.rvm/gems/ruby-2.5.3@qfr5/gems/railties-5.1.6.2/lib/rails/commands/server/server_command.rb:40:in `start'
20:41:10 web.1     | 	from /home/qf/.rvm/gems/ruby-2.5.3@qfr5/gems/railties-5.1.6.2/lib/rails/commands/server/server_command.rb:135:in `block in perform'
20:41:10 web.1     | 	from /home/qf/.rvm/gems/ruby-2.5.3@qfr5/gems/railties-5.1.6.2/lib/rails/commands/server/server_command.rb:130:in `tap'
20:41:10 web.1     | 	from /home/qf/.rvm/gems/ruby-2.5.3@qfr5/gems/railties-5.1.6.2/lib/rails/commands/server/server_command.rb:130:in `perform'
20:41:10 web.1     | 	from /home/qf/.rvm/gems/ruby-2.5.3@qfr5/gems/thor-0.20.3/lib/thor/command.rb:27:in `run'
20:41:10 web.1     | 	from /home/qf/.rvm/gems/ruby-2.5.3@qfr5/gems/thor-0.20.3/lib/thor/invocation.rb:126:in `invoke_command'
20:41:10 web.1     | 	from /home/qf/.rvm/gems/ruby-2.5.3@qfr5/gems/thor-0.20.3/lib/thor.rb:387:in `dispatch'
20:41:10 web.1     | 	from /home/qf/.rvm/gems/ruby-2.5.3@qfr5/gems/railties-5.1.6.2/lib/rails/command/base.rb:63:in `perform'
20:41:10 web.1     | 	from /home/qf/.rvm/gems/ruby-2.5.3@qfr5/gems/railties-5.1.6.2/lib/rails/command.rb:44:in `invoke'
20:41:10 web.1     | 	from /home/qf/.rvm/gems/ruby-2.5.3@qfr5/gems/railties-5.1.6.2/lib/rails/commands.rb:16:in `<top (required)>'
20:41:10 web.1     | 	from bin/rails:4:in `require'
20:41:10 web.1     | 	from bin/rails:4:in `<main>'
20:41:10 web.1     | exited with code 1
20:41:10 system    | sending SIGTERM to all processes
20:41:10 webpack.1 | terminated by SIGTERM
```

After some trial and error, I was able to narrow this down.

```
    - source: app
      target: /code/app
    - source: bin
      target: /code/bin
    - source: config
      target: /code/config
    - source: lib
      target: /code/lib
    - source: public
      target: /code/public
```

This didn't prevent creation of `/code/tmp` (the second backtrace), but did prevent the Webpack `/code/public` dir from being created. Commenting out the sync of `/code/public` prevented that error.

Therefore I believe that rsync was changing the owner or permissions of these directories in a way that was rendering my container un-runnable.

These are two "independent" issues but taken together, if you can't sync an entire directory (because of permissions issues), you also can't sync single files from that directory (because of command syntax appending a trailing slash). So in my case, for example, there appears to be no direct way to sync `Procfile.dev` without breaking the app's startup behavior due to permissions changes.

### Expected behavior

Sync supports single files, globs, or directories; file owner and permissions will be preserved if possible

### Your environment

```
[eronarn@ip-192-168-10-243 qf]$ garden version && kubectl version && docker version
0.10.2
Client Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.3", GitCommit:"5e53fd6bc17c0dec8434817e69b04a25d8ae0ff0", GitTreeState:"clean", BuildDate:"2019-06-06T01:44:30Z", GoVersion:"go1.12.5", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.7", GitCommit:"4683545293d792934a7a7e12f2cc47d20b2dd01b", GitTreeState:"clean", BuildDate:"2019-06-06T01:39:30Z", GoVersion:"go1.11.5", Compiler:"gc", Platform:"linux/amd64"}
Client: Docker Engine - Community
 Version:           19.03.0-rc2
 API version:       1.40
 Go version:        go1.12.5
 Git commit:        f97efcc
 Built:             Wed Jun  5 01:37:53 2019
 OS/Arch:           darwin/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.0-rc2
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.5
  Git commit:       f97efcc
  Built:            Wed Jun  5 01:42:10 2019
  OS/Arch:          linux/amd64
  Experimental:     true
 containerd:
  Version:          v1.2.6
  GitCommit:        894b81a4b802e4eb2a91d1ce216b8817763c29fb
 runc:
  Version:          1.0.0-rc8
  GitCommit:        425e105d5a03fabd737a126ad93d62a9eeede87f
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```

**Comment by edvald** [comment]
There is an existing issue for the single-file issue: #652  

We're generally pondering whether to stick with the rsync-style syntax, since it has a strong tendency to confuse users, but we'll likely tackle that matter separately. But this is helpful information. In fact I'll close the other issue in favor of this one, because this has more useful info (thanks for that btw!).

**Comment by elliott-davis** [comment]
Just chiming in to say we hit this today as well. Not a rails app but the symfony cache dir by default is `777`. We needed to sync the composer.json over to quickly update dependencies but syncing the entire git directory was clobbering permissions/users set during the container build. My workaround was to hot-reload as few directories as possible, then create a script that kicks off a `kubectl cp` before performing whatever action I needed to do.
